sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNullAwareHashJoinExec.scala

```diff
@@ -0,0 +1,205 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.joins
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, ExprCode, GenerateUnsafeProjection}
+import org.apache.spark.sql.catalyst.optimizer.{BuildRight, BuildSide}
+import org.apache.spark.sql.catalyst.plans._
+import org.apache.spark.sql.catalyst.plans.physical._
+import org.apache.spark.sql.execution.{CodegenSupport, ExplainUtils, SparkPlan}
+import org.apache.spark.sql.execution.metric.SQLMetrics
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.LongType
+
+case class BroadcastNullAwareHashJoinExec(
+    leftKeys: Seq[Expression],
+    rightKeys: Seq[Expression],
+    left: SparkPlan,
+    right: SparkPlan,
+    buildSide: BuildSide,
+    joinType: JoinType,
+    condition: Option[Expression]) extends BaseJoinExec with CodegenSupport {
+
+  // TODO support multi column NULL-aware anti join in future.
+  // See. http://www.vldb.org/pvldb/vol2/vldb09-423.pdf Section 6
+  // multi-column null aware anti join is much more complicated than single column ones.
+  require(leftKeys.length == 1, "leftKeys length should be 1")
+  require(rightKeys.length == 1, "rightKeys length should be 1")
+  require(right.output.length == 1, "null aware anti join only supports single column.")
+  require(joinType == LeftAnti, "joinType must be LeftAnti.")
+  require(buildSide == BuildRight, "buildSide must be BuildRight.")
+  require(SQLConf.get.nullAwareAntiJoinOptimizeEnabled,
+    "nullAwareAntiJoinOptimizeEnabled must be on for null aware anti join optimize.")
+
+  override lazy val metrics = Map(
+    "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"))
+
+  private val (streamed, broadcast) = (left, right)
+
+  override def simpleStringWithNodeId(): String = {
+    val opId = ExplainUtils.getOpId(this)
+    s"$nodeName $joinType ${buildSide} ($opId)".trim
+  }
+
+  override def requiredChildDistribution: Seq[Distribution] = {
+    UnspecifiedDistribution :: BroadcastDistribution(IdentityBroadcastMode) :: Nil
+  }
+
+  private[this] def genResultProjection: UnsafeProjection = {
+    UnsafeProjection.create(output, output)
+  }
+
+  override def output: Seq[Attribute] = {
+    left.output
+  }
+
+  private def prepareBroadcastHashedRelation = {
+    val buildSideRows = broadcast.executeBroadcast[Array[InternalRow]]().value
+    HashedRelation(buildSideRows.iterator,
+      BindReferences.bindReferences[Expression](
+        Seq(right.output.head), AttributeSeq(right.output)),
+      buildSideRows.length)
+  }
```

Comment:
why don't just call `buildPlan.executeBroadcast[HashedRelation]()` like BroadcastHashJoinExec?

Reply:
I keep this Array[InternalRow] originally for the support for multi-column,
but @agrawaldevesh said, maybe multi-column would be a goal that never reached, 
I will change back to HashedRelation, and in the future, i can always change it back.

let me do the update.

Reply:
done